### PR TITLE
Python Bridge

### PR DESCRIPTION
**Pull request by mackaiver**
We can call python methods from the streams xml now.
It works like this:
```
        <streams.PythonContext url="file:./python/processors/test.py">

            <streams.PythonProcessor method="process" />

        </streams.PythonContext>
```

I have not provided an abstraction on the python side of things. And I'm not sure if its necessary. The script has to look like this:
```
import Pyro4


@Pyro4.expose
class Processor(object):
    # add whatever methods you like
    def process(self, item):
        return item

    def add_thing_to_item(self, item):
        item['thing'] = 'an awesome thing'
        return item

    def add(self, a, b):
        return a + b


# the lower part has to look like this.
def main():
    Pyro4.Daemon.serveSimple(
            {
                Processor: 'streams.processors'
            },
            ns=True
    )

if __name__ == '__main__':
    main()

```

### CODE REVIEW

**Comment by mackaiver** [comment]
I will now start working on fixing the build system to work with python.

**Comment by alexeyegorov** [comment]
Current master branch merged. It looks like we can merge this pull request! @mackaiver

File: src/main/java/streams/PythonBridge.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+
+            String[] command = {"python", "-u", pathToPythonScript}; // -u for unbuffered python output
+
+            pythonProcess = new ProcessBuilder(command).redirectErrorStream(true).start();
+            stdin = new BufferedReader(new InputStreamReader(pythonProcess.getInputStream()));
+            stderr = new BufferedReader(new InputStreamReader(pythonProcess.getErrorStream()));
+
+            boolean isPyroDaemonRunning = false;
+            for (int i = 0; i < 50; i++) {
```
Comment (alexeyegorov) [inline src/main/java/streams/PythonBridge.java]: why 50?

File: src/main/java/streams/PythonBridge.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            throw new IOException("Python file not readable");
+        }
+        try {
+            String[] nameServerCommand = {"python", "-u", "-m", "Pyro4.naming"}; // -u for unbuffered python output
+
+            nameServerProcess = new ProcessBuilder(nameServerCommand).start();
+            BufferedReader stdin = new BufferedReader(new InputStreamReader(nameServerProcess.getInputStream()));
+            BufferedReader stderr = new BufferedReader(new InputStreamReader(nameServerProcess.getErrorStream()));
+            boolean isNameServerRunning = false;
+            for (int i = 0; i < 50; i++) {
```
Comment (alexeyegorov) [inline src/main/java/streams/PythonBridge.java]: why 50?

File: src/main/java/streams/PythonProcessor.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    @Override
+    public Data process(Data item) {
+        try {
+            @SuppressWarnings("unchecked") HashMap<String, Serializable> map = (HashMap) bridge.callMethod(method, item);
+            item.clear();
+            item.putAll(map);
+
+        } catch (IOException e) {
+            e.printStackTrace();
+            throw new RuntimeException("Python call failed.");
```
Comment (alexeyegorov) [inline src/main/java/streams/PythonProcessor.java]: better exception handling?

File: src/main/java/streams/PythonBridge.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+     * Calls a Python method by name.
+     * This methods delegates to the {@link PyroProxy#call} method.
+     * There is some more code in here which logs python output to the appropriate java log.
+     *
+     * @param name the name of the python method to call
+     * @param args the argumeents to pass to python
+     * @return the return value of the python method. The object has to be cast back to the proper type.
+     * @throws IOException in case an error occurs when calling the python method.
+     */
+    public Object callMethod(String name, Object... args) throws IOException {
```
Comment (alexeyegorov) [inline src/main/java/streams/PythonBridge.java]: this method receives DataItem, right? Shouldn't it also return such a hashmap directly? Or was your intention to enable all possible input/output combinations?

**Pull request merged**